src/GafferScene/InteractiveRender.cpp

```diff
@@ -217,6 +286,437 @@ void InteractiveRender::parentChanged( GraphComponent *child, GraphComponent *ol
+			for( std::vector<IECore::InternedString>::const_iterator it = childNames.begin(), eIt = childNames.end(); it != eIt; ++it )
+			{
+				SceneGraph *child = new SceneGraph();
+				child->m_name = *it;
+				child->m_parent = m_sceneGraph;
+				children.push_back( child );
+			}
+
+			// spawn child tasks:
+			set_ref_count( 1 + children.size() );
+			ScenePlug::ScenePath childPath = m_scenePath;
+			childPath.push_back( IECore::InternedString() ); // space for the child name
+			for( std::vector<SceneGraph *>::const_iterator it = children.begin(), eIt = children.end(); it != eIt; ++it )
+			{
+				childPath.back() = (*it)->m_name;
+				SceneGraphBuildTask *t = new( allocate_child() ) SceneGraphBuildTask(
+					m_scene,
+					m_context,
+					(*it),
+					childPath
+				);
+				
+				spawn( *t );
+			}
+
+			wait_for_all();
+			
+			// add visible children to m_sceneGraph->m_children:
+			for( std::vector<SceneGraph *>::const_iterator it = children.begin(), eIt = children.end(); it != eIt; ++it )
+			{
+				const BoolData *visibilityData = (*it)->m_attributes->member<BoolData>( SceneInterface::visibilityName );
+				if( visibilityData && !visibilityData->readable() )
+				{
+					continue;
+				}
+				m_sceneGraph->m_children.push_back( *it );
+			}
+			
+			return NULL;
+		}
+
+	private :
+
+		const ScenePlug *m_scene;
+		const Context *m_context;
+		SceneGraph *m_sceneGraph;
+		ScenePlug::ScenePath m_scenePath;
+};
+
+
+//////////////////////////////////////////////////////////////////////////
+// SceneGraphIteratorFilter implementation
+//
+// Does a serial, depth first traversal of a SceneGraph hierarchy based at
+// "start", and spits out SceneGraph* tokens:
+//
+//////////////////////////////////////////////////////////////////////////
+
+class InteractiveRender::SceneGraphIteratorFilter : public tbb::filter
+{
+	public:
+		SceneGraphIteratorFilter( InteractiveRender::SceneGraph *start ) :
+			tbb::filter( tbb::filter::serial_in_order ), m_current( start )
+		{
+			m_childIndices.push_back( 0 );
+		}
+		
+		void next()
+		{
+			// go down one level in the hierarchy if we can:
+			if( m_current->m_children.size() )
+			{
+				m_current = m_current->m_children[0];
+				m_childIndices.push_back(0);
+				return;
+			}
+			
+			while( m_childIndices.size() )
+			{
+				
+				// increment child index:
+				++m_childIndices.back();
+				
+				// find parent's child count - for the root we define this as 1:
+				size_t parentNumChildren = m_current->m_parent ? m_current->m_parent->m_children.size() : 1;
+				
+				if( m_childIndices.back() == parentNumChildren )
+				{
+					// we've got to the end of the child list, jump up one level:
+					m_childIndices.pop_back();
+					m_current = m_current->m_parent;
+					continue;
+				}
+				else
+				{
+					// move to next child of the parent:
+					m_current = m_current->m_parent->m_children[ m_childIndices.back() ];
+					return;
+				}
+			}
+		}
+		
+		void *operator()( void *item )
```

Comment:
As I understand it, this is overriding the pure virtual method defined on the filter base class? If so, please prefix it with `virtual` to make it clearer what is happening.